File: C#/Stacks/CircularStack/CircularStack.cs
/*******************************************************
 *  CircularStack.cs
 *  Created by Stephen Hall on 11/20/17.
 *  Copyright (c) 2017 Stephen Hall. All rights reserved.
 *  A Circular Stack implementation in C#
 ********************************************************/

namespace DataStructures.Stacks.CircularStack
{ 
    /// <summary>
    /// Circular Stack Class
    /// </summary>
    /// <typeparam name="T">Generic type</typeparam>
    public class CircularStack<T>
    {
        /// <summary>
        /// Private Members
        /// </summary>
        private T[] _array;
        private int _count;
        private int _size;
        private int _zeroIndex;

        /// <summary>
        /// Default Constructor
        /// </summary>
        public CircularStack()
        {
            _array = new T[_size = 10];
            _count = _zeroIndex = 0;
        }

        /// <summary>
        /// Circular Stack Constructor
        /// </summary>
        /// <param name="size">Size to initialize the stack to</param>
        public CircularStack(int size)
        {
            _array = new T[_size = size];
            _count = _zeroIndex = 0;
        }

        /// <summary>
        /// Pushes given data onto the stack if space is available
        /// </summary>
        /// <param name="data">Data to be added to the stack</param>
        /// <returns>item added to the stack</returns>
        public T Push(T data)
        {
            if (!IsFull())
            {
                _array[(_zeroIndex + _count) % _size] = data;
                _count++;
                return Top();
            }
            return default(T);
        }

        /// <summary>
        /// Pops item off the stack
        /// </summary>
        /// <returns>item popped off of the stack</returns>
        public T Pop()
        {
            if (IsEmpty())
                return default(T);
            T data = _array[(_count + _zeroIndex % _size) - 1];
            _array[(_count + _zeroIndex % _size) - 1] = _array[_zeroIndex];
            _array[_zeroIndex] = default(T);
            _count--;
            _zeroIndex = (_zeroIndex + 1) % _size;
            return data;
        }

        /// <summary>
        /// Gets the item onto of the stack
        /// </summary>
        /// <returns>item on top of the stack</returns>
        public T Top() => _array[((_zeroIndex + _count) % _size) - 1];

        /// <summary>
        /// Returns a value indicating if the stack is empty
        /// </summary>
        /// <returns>true if empty, false if not</returns>
        public bool IsEmpty() => _count == 0;

        /// <summary>
        /// Returns a value indicating if the stack is full
        /// </summary>
        /// <returns>True if full, false if not</returns>
        public bool IsFull() => _count == _size;
    }
    
}

File: Kotlin/Stacks/CircularStack/CircularStack.kt
/*******************************************************
 * CircularStack.kt
 * Created by Stephen Hall on 11/08/18.
 * Copyright (c) 2018 Stephen Hall. All rights reserved.
 * A Circular Stack implementation in Kotlin
 *******************************************************/

/**
 * Circular Stack Class
 * @param <T> Generic type
</T> */
class CircularStack<T>
/**
 * Circular Stack Constructor
 * @param size Size to initialize the stack to
 */
(size: Int) {
    /**
     * Private Members
     */
    private val array: Array<T?> = arrayOfNulls<Any>(size) as Array<T?>
    private var count: Int = 0
    private var size: Int = 0
    private var zeroIndex: Int = 0

    /**
     * Returns a value indicating if the stack is empty
     * @return true if empty, false if not
     */
    val isEmpty: Boolean
        get() = count == 0

    /**
     * Returns a value indicating if the stack is full
     * @return True if full, false if not
     */
    val isFull: Boolean
        get() = count == size

    /**
     * Default Constructor
     */
    constructor() : this(10) {}

    init {
        zeroIndex = 0
        count = zeroIndex
    }

    /**
     * Pushes given data onto the stack if space is available
     * @param data Data to be added to the stack
     * @return item added to the stack
     */
    fun push(data: T): T? {
        if (!isFull) {
            array[(zeroIndex + count) % size] = data
            count++
            return top()
        }
        return null
    }

    /**
     * Pops item off the stack
     * @return item popped off of the stack
     */
    fun pop(): T? {
        if (isEmpty)
            return null
        val data = array[count + zeroIndex % size - 1]
        array[count + zeroIndex % size - 1] = array[zeroIndex]
        array[zeroIndex] = null
        count--
        zeroIndex = (zeroIndex + 1) % size
        return data
    }

    /**
     * Gets the item onto of the stack
     * @return item on top of the stack
     */
    fun top(): T? {
        return array[(zeroIndex + count) % size - 1]
    }
}
